anvil/src/eth/otterscan/types.rs

```diff
@@ -0,0 +1,311 @@
+use ethers::types::{
+    Action, Address, Block, Bytes, Call, CallType, Create, CreateResult, Res, Suicide, Trace,
+    Transaction, TransactionReceipt, H256, U256,
+};
+use futures::future::join_all;
+use serde::{de::DeserializeOwned, Serialize};
+
+use crate::eth::{
+    backend::mem::Backend,
+    error::{BlockchainError, Result},
+};
+
+/// Patched Block struct, to include the additional `transactionCount` field expected by Otterscan
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase", bound = "TX: Serialize + DeserializeOwned")]
+pub struct OtsBlock<TX> {
+    #[serde(flatten)]
+    pub block: Block<TX>,
+    pub transaction_count: usize,
+}
+
+/// Block structure with additional details regarding fees and issuance
+#[derive(Serialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct OtsBlockDetails {
+    pub block: OtsBlock<Transaction>,
+    pub total_fees: U256,
+    pub issuance: Issuance,
+}
+
+/// Issuance information for a block. Expected by Otterscan in ots_getBlockDetails calls
+#[derive(Debug, Serialize, Default)]
+pub struct Issuance {
+    block_reward: U256,
+    uncle_reward: U256,
+    issuance: U256,
+}
+
+/// Holds both transactions and receipts for a block
+#[derive(Serialize, Debug)]
+pub struct OtsBlockTransactions {
+    pub fullblock: OtsBlock<Transaction>,
+    pub receipts: Vec<TransactionReceipt>,
+}
+
+/// Patched Receipt struct, to include the additional `timestamp` field expected by Otterscan
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct OtsTransactionReceipt {
+    #[serde(flatten)]
+    receipt: TransactionReceipt,
+    timestamp: u64,
+}
+
+/// Information about the creator address and transaction for a contract
+#[derive(Serialize, Debug)]
+pub struct OtsContractCreator {
+    pub hash: H256,
+    pub creator: Address,
+}
+
+/// Paginated search results of an account's history
+#[derive(Serialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct OtsSearchTransactions {
+    pub txs: Vec<Transaction>,
+    pub receipts: Vec<OtsTransactionReceipt>,
+    pub first_page: bool,
+    pub last_page: bool,
+}
+
+/// Otterscan format for listing relevant internal operations
+#[derive(Serialize, Debug, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct OtsInternalOperation {
+    pub r#type: OtsInternalOperationType,
+    pub from: Address,
+    pub to: Address,
+    pub value: U256,
+}
+
+/// Types of internal operations recognized by Otterscan
+#[derive(Serialize, Debug, PartialEq)]
+pub enum OtsInternalOperationType {
+    Transfer = 0,
+    SelfDestruct = 1,
+    Create = 2,
+    // The spec asks for a Create2 entry as well, but we don't have that info
+}
+
+#[derive(Serialize, Debug, PartialEq)]
```

Comment:
Let's document these as well! 😄

Reply:
completely forgot about the docs, tbh. I'll get on those tomorrow for everything :)

Reply:
done